platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -125,10 +126,15 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
                 else -> error("Unexpected configuration cache fingerprint: $input")
             }
         }
-        return if (firstReason == null) {
-            CheckedFingerprint.Valid
-        } else {
-            CheckedFingerprint.ProjectsInvalid(firstReason!!, projects.entries.filter { it.value.isInvalid }.map { it.key }.toSet())
+        return when (val path = firstInvalidatedPath) {
+            null -> CheckedFingerprint.Valid
+            else -> {
+                val firstInvalidatedState = projects.remove(path)!!
```

Comment:
🤔  The logic here was already overly complex, and I think we should work to simplify it.

I understand the desire to statically force the presence of the first invalidation reason, but the resulting code performs the same "dance on both sides of the fence". At the same time, we don't actually care (at the moment) about the owner of the first invalidation.

I suggest we drop the `firstInvalidatedPath` variable and simply rely on the insertion order preserved by the `mutableMapOf`.

I'd simplify the code then to something like:
```
        val invalidProjects = projects.filterValues { it.isInvalid }.mapValues { it.value.invalidationReason }
        return when {
            invalidProjects.isEmpty() -> CheckedFingerprint.Valid
            else -> CheckedFingerprint.ProjectsInvalid(invalidProjects.keys.first(), invalidProjects)
        }
```

And update the `ProjectsInvalid` to take the reason as the first constructor argument, and *all* the invalidations as the second argument. That should also simplify the implementation of that class.